---
title: "Create Matrices"
output: html_notebook
---

Run `1_wrangle_data.Rmd` to wrangle UMI matrices into cellphylo format (Fig. S1, step 1).

This notebook walks you through Fig. S1, steps 2-7, of the cellphylo workflow created for Mah & Dunn 2023. 

The steps are:  
2. Normalization and 3. Within-species batch integration  
4. Randomly subset matrices to 10 cells per cell type group  
5. Combine all 5 species into a single matrix based on gene orthology  
6. Cross-species integration  
7. PCA: rotation of data  

This will produe a 919 cell 919 PC matrix. This matrix is stored in the R object: xxx.  


# Preliminaries
```{r setup}
#set working directory
setwd("~/repo/cellphylo/scripts/cellphylo_analysis/")
```


# Steps 2 & 3: Normalization and Within-Species Integration  
For each species, we use the Seurat workflow to normalize counts using a negative binomial model (via SCTransform) and integrate sequencing batches within each species.    

Parameter values used by Mah & Dunn (2023) for `within_species_integration` are set as default:  

```{r}

#human
human_obj <- within_species_integration(path_to_matrix = "matrix/wrangled/matrix_wrangled_humanF", n_features=3000, k_weight = 100, print=TRUE)

#macF
macF_obj <- within_species_integration(path_to_matrix = "matrix/wrangled/matrix_wrangled_macF", n_features=3000, k_weight = 100, print=TRUE)

#macM
macM_obj <- within_species_integration(path_to_matrix = "matrix/wrangled/matrix_wrangled_macM", n_features=3000, k_weight = 100, print=TRUE)

#mouse
mouse_obj <- within_species_integration(path_to_matrix = "matrix/wrangled/matrix_wrangled_mouse", n_features=3000, k_weight = 100, print=TRUE)

#pig
pig_obj <- within_species_integration(path_to_matrix = "matrix/wrangled/matrix_wrangled_pigF", n_features=3000, k_weight = 100, print=TRUE)

```
