---
title: "Create Base Matrix"
output: html_notebook
---

Run `1_wrangle_data.Rmd` to wrangle UMI matrices into cellphylo format (Fig. S1, step 1).

This notebook walks you through Fig. S1, steps 2-7, of the cellphylo workflow created for Mah & Dunn 2023. 

The steps are:  
2 & 3. Normalization and within-species integration
4. Randomly subset matrices
5. Combine matrices by gene orthology 
6. Cross-species integration  
7. PCA: rotation and rank reduction of data  

This will produe a 919 cell 919 PC matrix from which the experiments of this paper were performed upon. This matrix is stored in the R object: xxx.  


# Preliminaries
```{r setup}
#set working directory
setwd("~/repo/cellphylo/scripts/cellphylo_analysis/")

#load libraries
library(tidyverse)
library(Seurat)
```


# Within-Species Analysis
The within-species analysis is performed separately for each species. 

## Steps 2 & 3. Normalization and Within-species Integration  
Fig. S1 steps 2 and 3

For each species, we use the Seurat workflow to normalize counts using a negative binomial model (via SCTransform) and integrate sequencing batches within each species.    

This step is memory intensive and you may have to run this on an HPC cluster.  
The human data set required 125 GiB RAM per CPU.  
All other species' data sets required 50 Gib ram per CPU.

Parameter values used by Mah & Dunn (2023) for `within_species_integration` are set as default.  

```{r}

#human
human_obj <- within_species_integration(path_to_matrix = "matrix/wrangled/matrix_wrangled_humanF", n_features=3000, k_weight = 100, print=TRUE)

#macF
macF_obj <- within_species_integration(path_to_matrix = "matrix/wrangled/matrix_wrangled_macF", n_features=3000, k_weight = 100, print=TRUE)

#macM
macM_obj <- within_species_integration(path_to_matrix = "matrix/wrangled/matrix_wrangled_macM", n_features=3000, k_weight = 100, print=TRUE)

#mouse
mouse_obj <- within_species_integration(path_to_matrix = "matrix/wrangled/matrix_wrangled_mouse", n_features=3000, k_weight = 100, print=TRUE)

#pig
pig_obj <- within_species_integration(path_to_matrix = "matrix/wrangled/matrix_wrangled_pigF", n_features=3000, k_weight = 100, print=TRUE)

```

Moving forward, we want to use the matrix of integrated Pearson's residuals, labelled: `matrix_sct_integrated_pearson_residuals`.  

## Step 4. Randomly subset matrices  
Fig. S1 step 4

### Identify the cell type with the fewest replicates
To make later analyses computationally feasible, we randomly subset each species' matrix so that all cell type groups across all five species have the same number of replicates. This means subsetting to the smallest number of replicates displayed by any one cell type group.  

```{r}

count_replicates("matrix/within-species")

```

We determine that the cell type group with the smallest number of replicates are mouse B cells, with 10 cells.  

### Subset the matrices 

For each species, randomly subset the full matrix to 10 cells per cell type group.  
```{r}
#human
subset_cells(path_to_matrix = "matrix/within-species/matrix_sct_integrated_pearson_residuals_human", n_reps = 10, print=TRUE)
#macF
subset_cells(path_to_matrix = "matrix/within-species/matrix_sct_integrated_pearson_residuals_macF", n_reps = 10, print=TRUE)
#macM
subset_cells(path_to_matrix = "matrix/within-species/matrix_sct_integrated_pearson_residuals_macM", n_reps = 10, print=TRUE)
#mouse
subset_cells(path_to_matrix = "matrix/within-species/matrix_sct_integrated_pearson_residuals_mouse", n_reps = 10, print=TRUE)
#pig
subset_cells(path_to_matrix = "matrix/within-species/matrix_sct_integrated_pearson_residuals_pig", n_reps = 10, print=TRUE)

```

# Cross-species Analysis
The cross-species analysis is performed on a combined matrix of all five species (Fig. S1, steps 5-7).

## Step 5. Combine matrices by gene orthology 
Fig. S1 step 5

We will combine the species matrix into a single matrix based on shared orthlogs.  

### Wrangle the orthlogs table  

Gene symbols vary between species. We need to create a mapping between gene symbols before we can combine the species' matrices by gene orthology. We will obtain this mapping from [Ensembl Biomart](https://www.ensembl.org/biomart/martview/).  

The below code is designed to data wrangle the Ensembl ortholog file downloaded for Mah and Dunn (2023). This wrangling facilitates downstream analysis. I have provided this orthology table: `Aqhumor_ensembl_orthologs.txt`. Put this orthology table in the `ann` folder. 


```{r}
#library(tidyverse)

# read in ensembl ontology file
ensembl_map <- read_tsv("ann/AqHumor_ensembl_orthologs.txt")
# standardize column names manually
colnames(ensembl_map) <- c("human_genename", "macM_genename","macM_homologytype", "macF_genename", "macF_homologytype","mouse_genename","mouse_homologytype", "pig_genename", "pig_homologytype")
#remove NAs
ensembl_map_noNA <- na.omit(ensembl_map)
#filter for lines where every seq is a one2one ortholog
ensembl_map_one2one <- filter(ensembl_map_noNA, macM_homologytype == "ortholog_one2one" & macF_homologytype == "ortholog_one2one" & mouse_homologytype == "ortholog_one2one" & pig_homologytype == "ortholog_one2one")
#add gene identifiers
ensembl_map_final <- mutate(ensembl_map_one2one, gene_identifier=paste0("gene_",1:nrow(ensembl_map_one2one)))
#select just gene name columns
ensembl_map_final <- select(ensembl_map_final, gene_identifier, macM_genename, macF_genename, mouse_genename, pig_genename, human_genename)
colnames(ensembl_map_final) <- c("gene_identifier", "macM", "macF", "mouse", "pig", "human")
                         

write.table(ensembl_map_final, file = "ann/AqHumor_ensembl_orthologs_wrangled.tsv", quote=FALSE, sep="\t", col.names=TRUE, row.names=FALSE)

```

This produces `AqHumor_ensembl_orthologs_wrangled.tsv` and TSV table mapping gene symbols across species. We use the generic gene_identifier (gene-1, gene-2...) moving forward.

### Combine all five species' matrices by shared orthologs

Combine the subsetted matrices by shared orthlogs.  

```{r}

matrix_dir_path <- "matrix/subset/"

#pursue old order for now
matrix_list <- c(paste0(matrix_dir_path,"matrix_subset_10_rep_macM"),paste0(matrix_dir_path,"matrix_subset_10_rep_macF/"),paste0(matrix_dir_path,"matrix_subset_10_rep_mouse"),paste0(matrix_dir_path,"matrix_subset_10_rep_pig/"),paste0(matrix_dir_path,"matrix_subset_10_rep_human/"))

#matrix_list <- c(paste0(matrix_dir_path,"matrix_subset_10_rep_human"),paste0(matrix_dir_path,"matrix_subset_10_rep_macF/"),paste0(matrix_dir_path,"matrix_subset_10_rep_macM"),paste0(matrix_dir_path,"matrix_subset_10_rep_mouse/"),paste0(matrix_dir_path,"matrix_subset_10_rep_pig/"))


combined_matrix <- combine_matrices(matrix_list = matrix_list, ortholog_map_path = "ann/AqHumor_ensembl_orthologs_wrangled.tsv", print=TRUE)

```

This produces the matrix `matrix/combined_by_orthology/matrix_combined` combined matrix is not yet cross-species integrated.

## Step 6. Cross-Species integration
Fig. S1 step 6

Use the Seurat workflow to integrate counts across species. Default values for cellphylo::cross_species_integration are what was used for Mah & Dunn (2023).

```{r}

cross_species_int <- cross_species_integration(matrix_path = "matrix/combined_by_orthology/matrix_combined")

#let's visualize the cross-species integration
#check by species integrated?
p1 <- DimPlot(cross_species_int, reduction = "umap", group.by = "species_id")
p1
#check by cell types separate?
p2 <- DimPlot(cross_species_int, reduction = "umap", group.by = "cell_type_id")
p2
#cell types in each species present in the same clusters across species?
p3 <- DimPlot(cross_species_int, reduction = "umap", group.by = "cell_type_id", split.by = "species_id")
p3
```
This produces the matrix `mtrix/cross-species_integration/matrix_cross-species_integrated`.

#7. PCA: rotation and rank reduction of the data 
Fig. S1 step 7

## Inspect the distribution of variance across principal components
```{r}
eig_plots <- plot_eigs(matrix_path = "matrix/cross-species_integration/matrix_cross-species_integrated/", highlight=20, print=TRUE)

```

This produces four plots:
1_eigenvalue_vs_PC: plot of eigenvalues against PCs. PC 20 is indicated with a red dot for Mah & Dunn (2023)
2_eigenvalue_cumulative_sum: cumulative sum of eigenvalues plotted against PCs
3_eigenvalue_vs_variance: plot of eigenvalue vs percentage variance described by each PC.
4_eigenvalue plot of percenrtage variance described by each PC against PCs. PC 20 is indicated wiht a red dot for Mah & Dunn (2023.)

From these plots we can identify an elbow at PC20. 

## Use PCA to rotate and rank reduce the matrix

Perform PCA on the matrix. Mean-centre before PCA and normalize variance by standard deviation. 
```{r}
pca_matrix <- run_pca(matrix_path="matrix/cross-species_integration/matrix_cross-species_integrated", n_PCs=919, print=TRUE)
```


Congratulations! You have created the 919 cell 919 PC matrix from Fig. S1 step 7. ALl other experiments and figures are based on this matrix.

