---
title: "single cell jackknife analysis"
---

This notebook provides the workflow to calculate the single cell jackknife (scjackknife) scores for Figure 4.

#Preliminaries
```{r}
setwd("~/repo/cellphylo/scripts/cellphylo_analysis/")

```

# Create scjackknife files
# Make scjack samples
To create each replicate 54 cell scjack matrix, we will repeatably sample 1 cell per cell type group per species, for the 54 cell type groups we used to create the phylogeny in Figure 4. We will use a 540 cell matrix (10 cells per cell type group, 54 cell type groups) to draw cell labels from. Because PCA must be performed on the full 919 cell matrix, we will create a 919 cell 20 PC matrix (Fig. S1, step C1) and pull out the cell labels we obtained from subsampling the cell labels of the 450 cell matrix. 

We do this 500 times to create 500 resampled 54 cell matrices. Each matrix is transformed into a `contml infile` and contml is used to infer  tree from it using the same options as that used for the jumble trees.

#Create the 919 cell 20 PC matrix
Fig. S1, C1
```{r run-pca-20pc}
pca_matrix <- run_pca(matrix_path="~/repo/aqhumor/AqHum_5Species/aqhumor_final/matrix_final/integrated/cross_species/cross_species_integrated/aqhumor_cross_species_integrated_mtx/", n_PCs=20, print=TRUE)
```

input:
MD5 (aqhumor_cross_species_integrated_mtx/matrix.mtx.gz) = daa2aef295e2b7f07f80de669b8a3c1c
output:
MD5 (contml_919cell_subset_20PC_mtx/matrix.mtx.gz) = 4869abf3f078f17655e1bfe4697f8b98

#Make scjack samples

This function will create 500 resampled 54 cell matrices and print them out as 10x feature-barcode matrices, infiles, and the list of cell ids present in each matrix.

Use cellphylo
```{r create-scjack-sample}

# The matrix to subsample cells from. This is the 540 cell matrix (Fig. S1, step C4)
subsample_matrix_path = "~/repo/aqhumor/AqHum_5Species/aqhumor_final/matrix_final/integrated/cross_species/cross_species_integrated/540_aqhumor_cross_species_integrated_mtx/"
#919 cell normalized, within + cross-species integrated matrix, not subsetted with black-listed cell types removed
# The matrix to draw values from. This should be the 919 cell 20 PC matrix (Fig. S1, step C1)
full_matrix_path = "~/repo/aqhuedmor/AqHum_5Species/aqhumor_final/scjackknife/scjack_54_jumble_search_TBE/scjack_analysis/contml_919cell_subset_20PC_mtx/"
#number of scjack samples to make
n_samples = c(1:500)

#function calls: subset_combined_matrix(), run_pca(), and make_infile()
scjack_sample <- make_scjack_sample(n_samples = n_samples, subsample_matrix_path = subsample_matrix_path, full_matrix_path = full_matrix_path, file_name="540_TBE")

```
input:
MD5 (540_aqhumor_cross_species_integrated_mtx/matrix.mtx.gz) = d616301cc9fdda896ed4eac52268f64f
MD5 (contml_919cell_subset_20PC_mtx/matrix.mtx.gz) = 4869abf3f078f17655e1bfe4697f8b98

output:
MD5 (input_scjack_540_TBE.tar.gz) = a6f526a2163cdaec150e6fbec5e251f4

# Run contml on the scjack infiles

`contml` was run using the same options as for the jumble trees:
  C: continuous data
  J: jumble, 100x with random seed
  G: global search

We ran this on a hpc cluster using 64 vCPUs and running 50 `contml` commands in parallel at a time. This run takes several hours.  

The bash scripts used to perform the contml runs are available in xxx.

Template script: in_script_scjack_parallel_540_TBE.sh

```
#!/bin/bash

#in_script_scjack_parallel_540_TBE.sh

#run contml on infile_n
#must not have an 'infile' indirectory - will replace with infile_n
#must have an (empty) outfile and outtree so can set

n=ITERATION_INDEX

cd /contml

echo "Run: ${n}"

#create random seed for jumble
RAND_NUM=$(shuf -i 1-2000000000 -n 1)
random_seed=$(($RAND_NUM * 2 + 1))
echo "seed: ${random_seed}"

outfile_new="outfile_scjack_540_TBE_${n}_${random_seed}"
outtree_new="outtree_scjack_540_TBE_${n}_${random_seed}"

##CHANGE HERE
infile="./scjack_540_infiles/infile_scjackknife_540_final_${n}"

echo "outfile: ${outfile_new}"
echo "outtree: ${outtree_new}"
echo "infile: ${infile}"

printf "${infile}\nF\n${outfile_new}\nC\nJ\n${random_seed}\n100\nG\nY\nF\n${outtree_new}\n"| ./phylip/phylip-3.697/exe/contml
#test
#printf "${infile}\nF\n${outfile_new}\nC\nJ\n${random_seed}\n2\nG\nY\nF\n${outtree_new}\n"| ./phylip/phylip-3.697/exe/contml

mv ${outtree_new} ./outtrees_scjack_540_TBE/${outtree_new}.tre
mv ${outfile_new} ./outfiles_scjack_540_TBE/${outfile_new}.txt


echo "Run ${n} complete"
```
output: MD5 (in_script_scjack_parallel_540_TBE.sh) = 94e8392ca2fe400db843398258075705

Make copies of template script: make_reps_scjack_540_TBE.sh
```
#!/bin/bash

#make_reps_scjack_540_TBE

### Modified from script produced by ChatGPT Dec 9 2022.
### Prompt: "Write a bash script that iteratively reads in a text file, copies and renames it a new iterative file name, and replaces a string in the file with the iteration index. Make it do it 100 times."

# Set the starting iteration index
index=1

# Set the filename of the original file
filename="in_script_scjack_parallel_540_TBE.sh"

# Set the string to be replaced in the file
replace="ITERATION_INDEX"

# Set the prefix for the new iterative filenames
new_filename_prefix="script_scjack_540_TBE_"

# Set the suffix for the new iterative filenames
new_filename_suffix=".sh"

# Set the maximum number of iterations
max_iterations=500


# Iterate until the original file does not exist or the maximum number of iterations is reached
while [ -f $filename ] && [ $index -le $max_iterations ]
do
  # Copy and rename the file with the current iteration index
  #don't need to do this
  #cp $filename $new_filename_prefix$index$new_filename_suffix

  # Replace the string in the copied file with the iteration index
  #sed -i "s/$replace/$index/g" $new_filename_prefix$index$new_filename_suffix
  #sed command did not work. Remove -i flag, input old file, pipe sed output to new file.
  sed "s/$replace/$index/g" $filename > $new_filename_prefix$index$new_filename_suffix
  
  #set permissions
  chmod u+x $new_filename_prefix$index$new_filename_suffix
  
  # Increment the iteration index
  index=$((index+1))
  
done
```
input:
MD5 (in_script_scjack_parallel_540_TBE.sh) = 94e8392ca2fe400db843398258075705
MD5 (make_reps_scjack_540_TBE.sh) = 09d45b0e2d4c1280dea6a9eec29333dd

output:
MD5 (scjack_scripts.tar.gz) = dca27db26274f013d2e17773bef800ec

Script to run the scjackknife analysis on cluster: run_scjack_parallel_540_TBE.sh

```
#!/bin/bash

#run_scjack_parallel_540_TBE.sh

#run Jumble in parallel
#only want to set up directories once
#need 1 empty outfile and outtree so you can indicate indiividual names for output files

cd /contml

mkdir outtrees_scjack_540_TBE
mkdir outfiles_scjack_540_TBE/

#make an empty outfile and outtree so contml prompts for custom file names
touch outfile
touch outtree

#ensure that there is no 'infile'. Will throw up error if already no infile.
#rm infile

echo "scjack jumble search preliminaries complete"

#send in the command. Remove last `&`. #chatgpt

#for i in {1..50}; do printf "script_scjack_540_TBE_$i.sh & %.0s"; done

#run the script 50 times in parallel
#repeat 50 times 
script_scjack_540_TBE_1.sh & script_scjack_540_TBE_2.sh 

echo "Commands 1-50 complete"

echo "ALL COMMANDS COMPLETE!!!!!"

exit

```
output: MD5 (run_scjack_parallel_540_TBE.sh) = 1022eb1cb9be46d76e5dd5996b1df20a0

#Infer scjackknife trees on cluster
EC2:

Amazon Linux 2023 AMI
Amazon Linux 2023 AMI 2023.0.20230419.0 x86_64 HVM kernel-6.1
64-bit (x86), boot mode: uefi-preferred, AMI ID: ami-02396cdd13e9a1257

Instance: c5a.16xlarge: 64 vCPU, 128 GiB mem

VPC - required  Info

vpc-066eb8998a0cf415c (partner_dunn-vpc)
10.9.17.0/24

Subnet Info

subnet-015e679f07bca9a60
partner_dunn-private-us-east-1a

30 GiB root volume

Private IP: 10.9.17.45

Run:

```
tmux new -s scjack_540_TBE
sudo service docker start
sudo docker run --rm --name contml -v /home/ec2-user/contml:/contml -it dunnlab/ubuntu-dev:20.04 /bin/bash 

./run_scjack_parallel_540_TBE.sh |tee -a run_scjack_parallel_540_TBE.log
```

input:
MD5 (run_scjack_parallel_540_TBE.sh) = 1022eb1cb9be46d76e5dd5996b1df20a
MD5 (input_scjack_540_TBE.tar.gz) = fbd5ff0c082cddfec39e2b084149a589
output:
MD5 (output_scjack_540_TBE.tar.gz) = 384d0762ca07dc73e349d65dbab72127

# Process output files
Some `contml` runs may have errored out, producing empty `outtree` files. Remove these files

Delete any empty tree files
```
find ./ -size 0 -print -delete
```

Copy `outfiles` that are not empty into a new folder

```
find . -type f -print| xargs grep "Ln Likelihood"| cut -d ":" -f 1| xargs -I '{}' cp '{}' ../outfiles_scjack_540_TBE_processed/
```

439 trees remaining

input:
MD5 (output_scjack_540_TBE.tar.gz) = 384d0762ca07dc73e349d65dbab72127
output:
MD5 (outfiles_scjack_540_TBE_processed.tar.gz) = 54425c395958a7db2eb9a403a6e26c0f
MD5 (outtrees_scjack_540_TBE_processed.tar.gz) = 45113b16c94ede167c3ea5036ad48e0b


Removing files causes the index between the scjack matrices and `outtree` and `outfile` files to be mismatched, causing downstream analysis issues. Remove those scjack matrices that produced errored runs.
use cellphylo:

```{r id-errored-runs}
outtrees_dir = "~/repo/aqhumor/AqHum_5Species/aqhumor_final/scjackknife/scjack_54_jumble_search_TBE/scjack_run/output_scjack_540_TBE/outtrees_scjack_540_TBE/"
matrix_name_pattern = "scjackknife_540_TBE"

id_errored_runs(outtrees_dir = outtrees_dir, matrix_name_pattern = matrix_name_pattern)

```

input:
MD5 (../output_scjack_540_TBE.tar.gz) = 384d0762ca07dc73e349d65dbab72127
output:
MD5 (errored_runs.txt) = 530e1f2e531231b1f257e20fd8e3b674

Remove these matrices using bash:
```
cd scjack_matrices
rm -r $(<errored_runs.txt)

```
input: 
MD5 (errored_runs.txt) = 530e1f2e531231b1f257e20fd8e3b674
MD5 (input_scjack_540_TBE.tar.gz) = fbd5ff0c082cddfec39e2b084149a589

output:
MD5 (scjack_matrices_processed.tar.gz) = d6c528205438590a3e987ce0541b230d

# Calculate the scjack scores

# Add common labels to scjack trees
To compare scjack trees, we must convert their tip labels to one that is common across all scjack trees. 

```{r}
#how many trees do we have in the tree directory?
n_trees =439
matrix_dir_path = "~/repo/aqhumor/AqHum_5Species/aqhumor_final/scjackknife/scjack_54_jumble_search_TBE/scjack_run/output_scjack_540_TBE_processed/scjack_matrices_processed/"
tree_dir_path = "~/repo/aqhumor/AqHum_5Species/aqhumor_final/scjackknife/scjack_54_jumble_search_TBE/scjack_run/output_scjack_540_TBE_processed/outtrees_scjack_540_TBE_processed/"
file_name_prefix = "outtrees_scjack_540_TBE"

labelled_trees <- label_scjack_trees(n_trees = n_trees, matrix_dir_path = matrix_dir_path, tree_dir_path = tree_dir_path, print=TRUE, file_name_prefix=file_name_prefix)
```
input:
MD5 (outtrees_scjack_540_TBE_processed.tar.gz) = 45113b16c94ede167c3ea5036ad48e0b
MD5 (scjack_matrices_processed.tar.gz) = d6c528205438590a3e987ce0541b230d

output
MD5 (outtrees_scjack_540_TBE_common_label.tar.gz) = 078f68cd69e425505f2942e3f6ccfec8
MD5 (outtrees_scjack_540_TBE_label.tar.gz) = 6951bb0225d2c74c9583271b20db3844

## Add common labels to the focal tree for Fig. 4
Wrangle best jumble tree
```{r boot_files2}

tree <- read.tree("aqhumor_final/scjackknife/scjack_54_jumble_search_TBP/outtree_54_jumble_search_1249177875.tre")

labelled_tree <- label_tree(matrix_path = "aqhumor_final/jumble/54_jumble_search/contml_54cell_subset_pca_var_norm_20PC_mtx/", tree_path = "aqhumor_final/scjackknife/scjack_54_jumble_search_TBP/outtree_54_jumble_search_1249177875.tre", file_name = "outtree_54_jumble_search_1249177875_label.tre", print=TRUE)

#make common labels - code stolen from scjack label
  species <- sapply(strsplit(labelled_tree$tip.label, "_"), function(v){return(v[1])})
  cluster_id <- sapply(strsplit(labelled_tree$tip.label, "_"), function(v){return(v[3])})
  new_label <- paste0(species, "_", cluster_id)
  common_labels_tree <- labelled_tree
  common_labels_tree$tip.label <- new_label
  #print out common labels tree
    write.tree(common_labels_tree, file="outtree_54_jumble_search_1249177875_common_label.tre")

```

#Run booster
Booster expects a newick multiphylo file
```{r boot-files}

library(ape)

multi_trees <- combine_multi_trees("~/repo/aqhumor/AqHum_5Species/aqhumor_final/scjackknife/scjack_54_jumble_search_TBE/scjack_run/output_scjack_540_TBE_processed/outtrees_scjack_540_TBE_common_label/")


write.tree(multi_trees, file="multiphylo_scjack_540_TBE_common_label.tre")

```
input:
MD5 (outtrees_scjack_540_TBE_common_label.tar.gz) = 078f68cd69e425505f2942e3f6ccfec8

output:
MD5 (multiphylo_scjack_540_TBE_common_label.tre) = fa22d9c95d122f72a7177c2ba0fbdf55


#Calculate TBE scores for Fig. 4 focal phylogeny
#New TBE with original fig 4 best jumble
Try out the best jumble tree (fig. 4 tree)
Did not include tree in multiphylo
```
../../booster/booster_macos64 -i outtree_54_jumble_search_1249177875_common_label.tre -b multiphylo_scjack_540_TBE_common_label.tre  > outtree_54_jumble_search_1249177875_scored_with_scjack540TBE.tre
```
