---
title: "Create matrices"
---

Run `1_wrangle_data.Rmd` to wrangle UMI matrices into cellphylo format (Fig. S1, step 1).

This notebook walks you through Fig. S1, steps 2-6, of the cellphylo workflow created for Mah & Dunn 2023.

The steps are:  
2 & 3. Normalization and within-species integration
4. Randomly subset matrices
5. Combine matrices by gene orthology 
6. Cross-species integration  

# Preliminaries
```{r setup}
#set working directory to the analysis directory
setwd("~/repo/cellphylo/analysis")

#load libraries
library(tidyverse)
library(Seurat)
library(DropletUtils)
```


# Within-Species Analysis
The within-species analysis is performed separately for each species. 

## Steps 2 & 3. Normalization and Within-species Integration  
Fig. S1 steps 2 and 3

For each species, we use the Seurat workflow to normalize counts using a negative binomial model (via SCTransform) and integrate sequencing batches within each species.    

This step is memory intensive and you may have to run this on an hpc cluster.  
  The human data set required 125 GiB RAM per CPU.  
  All other species' data sets required 50 Gib ram per CPU.

```{r within-species-integration}

#human
human_obj <- within_species_integration(path_to_matrix = "matrix/wrangled/matrix_wrangled_human", n_features=3000, k_weight = 100, print=TRUE)

#macF
macF_obj <- within_species_integration(path_to_matrix = "matrix/wrangled/matrix_wrangled_macF", n_features=3000, k_weight = 100, print=TRUE)

#macM
macM_obj <- within_species_integration(path_to_matrix = "matrix/wrangled/matrix_wrangled_macM", n_features=3000, k_weight = 100, print=TRUE)

#mouse
mouse_obj <- within_species_integration(path_to_matrix = "matrix/wrangled/matrix_wrangled_mouse", n_features=3000, k_weight = 100, print=TRUE)

#pig
pig_obj <- within_species_integration(path_to_matrix = "matrix/wrangled/matrix_wrangled_pig", n_features=3000, k_weight = 100, print=TRUE)

```

Moving forward, we want to use the matrix of normalized and integrated Pearson's residuals that is produced after integration. Note: these matrices are too large to provide in the repo but they are available upon request. Downstream matrices calculated from these matrices are provided, and described below.

## Step 4. Randomly subset matrices  
Fig. S1 step 4

### Identify the cell type with the fewest replicates
To make later analyses computationally feasible, we randomly subset each species' matrix so that all cell type groups across all five species have the same number of replicates. This means subsetting to the smallest number of replicates displayed by any one cell type group.  

The function `count_replicates` counts the number of replicate cells per cell type group for each matrix.
```{r count-replicates}

#`matrix/within-species` is a directory containing the five within-species integrated matrices that were created above. These matrices are too large to provide in the repo but are available upon request.

count_replicates("matrix/within-species")

```

We determine that the cell type group with the smallest number of replicates are mouse B cells, with 10 replicate cells.


### Subset the matrices 

For each species, randomly subset the full matrix to 10 cells per cell type group.  

The normalized and within-species matrices are not provided due to size, but we provide the subsetted matrices in *`analysis/matrix/subset`*. These subsetted matrices are provided to ensure reproducibility, since subsetting is random. However, if you'd like to create your own subsetted matrices, use the code below.

```{r subset-cells} 


  
#human
subset_cells(path_to_matrix = "matrix/within-species_norm_and_integration/integrated_pearson_residuals/matrix_sct_integrated_pearson_residuals_human", n_reps = 10, print=TRUE)
#macF
subset_cells(path_to_matrix = "matrix/within-species_norm_and_integration/integrated_pearson_residuals/matrix_sct_integrated_pearson_residuals_macF", n_reps = 10, print=TRUE)
#macM
subset_cells(path_to_matrix = "matrix/within-species_norm_and_integration/integrated_pearson_residuals/matrix_sct_integrated_pearson_residuals_macM", n_reps = 10, print=TRUE)
#mouse
subset_cells(path_to_matrix = "matrix/within-species_norm_and_integration/integrated_pearson_residuals/matrix_sct_integrated_pearson_residuals_mouse", n_reps = 10, print=TRUE)
#pig
subset_cells(path_to_matrix = "matrix/within-species_norm_and_integration/integrated_pearson_residuals/matrix_sct_integrated_pearson_residuals_pig", n_reps = 10, print=TRUE)

```



# Cross-species Analysis
The cross-species analysis (steps 5 and 6) is performed on a combined matrix of all five species. 

## Step 5. Combine matrices by gene orthology 
Fig. S1 step 5

We will combine the species matrix into a single matrix based on shared orthologs.  

### Wrangle the orthlogs table  

Gene symbols vary between species. We need to create a mapping between gene symbols before we can combine the species' matrices by gene orthology. We will obtain this mapping from [Ensembl Biomart](https://www.ensembl.org/biomart/martview/). We provide the raw file we downloaded from Ensembl Biomart in `ann/Aqhumor_ensembl_orthologs.txt`.

The below code is designed to data wrangle the Ensembl ortholog file `Aqhumor_ensembl_orthologs.txt` for downstream analyses. This produces `AqHumor_ensembl_orthologs_wrangled.tsv`.


```{r wrangle-ortholog-map}
#library(tidyverse)

# read in ensembl ontology file
ensembl_map <- read_tsv("ann/AqHumor_ensembl_orthologs.txt")
# standardize column names manually
colnames(ensembl_map) <- c("human_genename", "macM_genename","macM_homologytype", "macF_genename", "macF_homologytype","mouse_genename","mouse_homologytype", "pig_genename", "pig_homologytype")
#remove NAs
ensembl_map_noNA <- na.omit(ensembl_map)
#filter for lines where every seq is a one2one ortholog
ensembl_map_one2one <- filter(ensembl_map_noNA, macM_homologytype == "ortholog_one2one" & macF_homologytype == "ortholog_one2one" & mouse_homologytype == "ortholog_one2one" & pig_homologytype == "ortholog_one2one")
#add gene identifiers
ensembl_map_final <- mutate(ensembl_map_one2one, gene_identifier=paste0("gene_",1:nrow(ensembl_map_one2one)))
#select just gene name columns
ensembl_map_final <- select(ensembl_map_final, gene_identifier, macM_genename, macF_genename, mouse_genename, pig_genename, human_genename)
colnames(ensembl_map_final) <- c("gene_identifier", "macM", "macF", "mouse", "pig", "human")
                         

write.table(ensembl_map_final, file = "ann/AqHumor_ensembl_orthologs_wrangled.tsv", quote=FALSE, sep="\t", col.names=TRUE, row.names=FALSE)

```

This produces `AqHumor_ensembl_orthologs_wrangled.tsv` and TSV table mapping gene symbols across species. We use the generic gene_identifier (gene-1, gene-2...) moving forward.

### Combine all five species' matrices by shared orthologs

Combine the subsetted matrices (Fig. S1, step 4) by shared orthologs into a single combined matrix (Fig. S1, step 5)

```{r combine-matrices}

matrix_dir_path <- "matrix/subset/"

#pursue old order for now
matrix_list <- c(paste0(matrix_dir_path,"integrated_subset_mtx_macM"),paste0(matrix_dir_path,"integrated_subset_mtx_macF"),paste0(matrix_dir_path,"integrated_subset_mtx_mouse"),paste0(matrix_dir_path,"integrated_subset_mtx_pig"),paste0(matrix_dir_path,"integrated_subset_mtx_human"))

combined_matrix <- combine_matrices(matrix_list = matrix_list, ortholog_map_path = "ann/AqHumor_ensembl_orthologs_wrangled.tsv", print=TRUE)

```

This produces the matrix `matrix/combined_by_orthology/aqhumor_combined_mtx/`.

`aqhumor_combined_mtx` is not yet cross-species integrated.


## Step 6. Cross-Species integration
Fig. S1 step 6

Use the Seurat workflow to integrate counts across species. Default values for cellphylo::cross_species_integration are what was used for Mah & Dunn (2023).

```{r cross-species-integration}

cross_species_int <- cross_species_integration(matrix_path = "matrix/combined_by_orthology/aqhumor_combined_mtx/")

#let's visualize the cross-species integration
#check by species integrated?
p1 <- DimPlot(cross_species_int, reduction = "umap", group.by = "species_id")
p1
#check by cell types separate?
p2 <- DimPlot(cross_species_int, reduction = "umap", group.by = "cell_type_id")
p2
#cell types in each species present in the same clusters across species?
p3 <- DimPlot(cross_species_int, reduction = "umap", group.by = "cell_type_id", split.by = "species_id")
p3
```
This produces the matrix `matrix/cross-species_integration/aqhumor_cross_species_integrated_mtx"` and also an R object `cross_species_integrated.rds` that is the Seurat object created by this workflow.

For convenience, we have provided this matrix in *`matrix/cross-species_integration/aqhumor_cross_species_integrated_mtx`*
