---
title: "single cell jackknife analysis"
---

This notebook provides the workflow to calculate the single cell jackknife (scjackknife) scores for Figure 4. The scjackknife procedure randomly draws replicate cells from a larger matrix to create the 54 cell matrix. We repeat this procedure 500 times and infer an `scjackknife tree` from each. These trees are used to calculate TBE (transfer bootstrap expectation) scores, which served as the `scjackknife scores` in Figure 4.

#Preliminaries
```{r}
setwd("~/repo/cellphylo/scripts/cellphylo_analysis/")

#load 
library(ape)

```

# scJackknife analysis
This workflow creates scjackknife matrices, infers a scjackknife tree from each, removes errored runs and calculates TBE scores from the scjackknife trees. 

# Make scjack samples
To create each replicate 54 cell scjack matrix, we will repeatably sample 1 cell per cell type group per species, for the 54 cell type groups we used to create the phylogeny in Figure 4. We will use a 540 cell matrix (10 cells per cell type group, 54 cell type groups) to draw cell labels from. Because PCA must be performed on the full 919 cell matrix, we will create a 919 cell 20 PC matrix (Fig. S1, step C1) and pull out the values corresponding to the cell labels we obtained from subsampling the 450 cell matrix. 

This in effect is the same as subsetting the 919 cell 20 PC matrix to 450 cell and 20 PCs (Fig. S1, step C4), but due to a quirk in the workflow the former was performed.

We do this 500 times to create 500 resampled 54 cell matrices. Each matrix is transformed into a `contml infile` and contml is used to infer  tree from it using the same settings as that used for the jumble trees.

This function will print out the 54 cell subsampled matrices as 10x feature-barcode matrices, infiles, and the list of cell ids present in each matrix.


```{r create-scjack-sample}

# The matrix to subsample cells from. This is the 540 cell matrix (Fig. S1, step C4)
subsample_matrix_path = "~/repo/aqhumor/AqHum_5Species/aqhumor_final/matrix_final/integrated/cross_species/cross_species_integrated/540_aqhumor_cross_species_integrated_mtx/"
# The matrix to draw values from. This should be the 919 cell 20 PC matrix (Fig. S1, step C1)
full_matrix_path = "~/repo/aqhuedmor/AqHum_5Species/aqhumor_final/scjackknife/scjack_54_jumble_search_TBE/scjack_analysis/contml_919cell_subset_20PC_mtx/"
#number of scjack samples to make
n_samples = c(1:500)

#function calls: subset_combined_matrix(), run_pca(), and make_infile()
scjack_sample <- make_scjack_sample(n_samples = n_samples, subsample_matrix_path = subsample_matrix_path, full_matrix_path = full_matrix_path, file_name="540_TBE")

```

# Run contml on the scjack infiles
Use the infiles created above to run `contml`.

`contml` was run using the same options as for the jumble trees:
  C: continuous data
  J: jumble, 100x with random seed
  G: global search

We ran this on a hpc cluster using 64 vCPUs and running 50 `contml` commands in parallel at a time. This took 4-5 hours.

Each scjack `contml` run requires its own bash script. We created and executed 500 bash scripts using the following workflwo:

## Template script: in_script_scjack_parallel_540_TBE.sh
We created a template for the script that is executed for each contml run.

```
#!/bin/bash

#in_script_scjack_parallel_540_TBE.sh

#run contml on infile_n
#must not have an 'infile' indirectory - will replace with infile_n
#must have an (empty) outfile and outtree so can set

n=ITERATION_INDEX

cd /contml

echo "Run: ${n}"

#create random seed for jumble
RAND_NUM=$(shuf -i 1-2000000000 -n 1)
random_seed=$(($RAND_NUM * 2 + 1))
echo "seed: ${random_seed}"

outfile_new="outfile_scjack_540_TBE_${n}_${random_seed}"
outtree_new="outtree_scjack_540_TBE_${n}_${random_seed}"

##CHANGE HERE
infile="./scjack_540_infiles/infile_scjackknife_540_final_${n}"

echo "outfile: ${outfile_new}"
echo "outtree: ${outtree_new}"
echo "infile: ${infile}"

printf "${infile}\nF\n${outfile_new}\nC\nJ\n${random_seed}\n100\nG\nY\nF\n${outtree_new}\n"| ./phylip/phylip-3.697/exe/contml
#test
#printf "${infile}\nF\n${outfile_new}\nC\nJ\n${random_seed}\n2\nG\nY\nF\n${outtree_new}\n"| ./phylip/phylip-3.697/exe/contml

mv ${outtree_new} ./outtrees_scjack_540_TBE/${outtree_new}.tre
mv ${outfile_new} ./outfiles_scjack_540_TBE/${outfile_new}.txt


echo "Run ${n} complete"
```

## Make copies of template script: make_reps_scjack_540_TBE.sh
We made 500 copies of the template script using the below script.

```
#!/bin/bash

#make_reps_scjack_540_TBE

### Modified from script produced by ChatGPT Dec 9 2022.
### Prompt: "Write a bash script that iteratively reads in a text file, copies and renames it a new iterative file name, and replaces a string in the file with the iteration index. Make it do it 100 times."

# Set the starting iteration index
index=1

# Set the filename of the original file
filename="in_script_scjack_parallel_540_TBE.sh"

# Set the string to be replaced in the file
replace="ITERATION_INDEX"

# Set the prefix for the new iterative filenames
new_filename_prefix="script_scjack_540_TBE_"

# Set the suffix for the new iterative filenames
new_filename_suffix=".sh"

# Set the maximum number of iterations
max_iterations=500


# Iterate until the original file does not exist or the maximum number of iterations is reached
while [ -f $filename ] && [ $index -le $max_iterations ]
do
  # Copy and rename the file with the current iteration index
  #don't need to do this
  #cp $filename $new_filename_prefix$index$new_filename_suffix

  # Replace the string in the copied file with the iteration index
  #sed -i "s/$replace/$index/g" $new_filename_prefix$index$new_filename_suffix
  #sed command did not work. Remove -i flag, input old file, pipe sed output to new file.
  sed "s/$replace/$index/g" $filename > $new_filename_prefix$index$new_filename_suffix
  
  #set permissions
  chmod u+x $new_filename_prefix$index$new_filename_suffix
  
  # Increment the iteration index
  index=$((index+1))
  
done
```

## Script to execute the scjackknife analysis on cluster: run_scjack_parallel_540_TBE.sh

The above workflow created:
- 500 infiles representing 500 resampled 54 cell matrices
- 500 contml scripts to execute a contml run for each infile

Now execute the scjackknife analysis on a cluster using the below script:
```
#!/bin/bash

#run_scjack_parallel_540_TBE.sh

#run Jumble in parallel
#only want to set up directories once
#need 1 empty outfile and outtree so you can indicate indiividual names for output files

cd /contml

mkdir outtrees_scjack_540_TBE
mkdir outfiles_scjack_540_TBE/

#make an empty outfile and outtree so contml prompts for custom file names
touch outfile
touch outtree

#ensure that there is no 'infile'. Will throw up error if already no infile.
#rm infile

echo "scjack jumble search preliminaries complete"

#send in the command. Remove last `&`. #chatgpt

#for i in {1..50}; do printf "script_scjack_540_TBE_$i.sh & %.0s"; done

#run the script 50 times in parallel. The below line only executes two scripts. Add all the template scripts you want to run in parallel to this command:
script_scjack_540_TBE_1.sh & script_scjack_540_TBE_2.sh 


exit

```
This produced 500 outtrees and outfiles representing a scjackknife tree inferred from each resampled 54 cell matrix.

# Process output files
Some `contml` runs may have errored out, producing empty `outtree` files. Remove these files

Delete any empty tree files
```
find ./ -size 0 -print -delete
```

Copy `outfiles` that are not empty into a new folder

```
find . -type f -print| xargs grep "Ln Likelihood"| cut -d ":" -f 1| xargs -I '{}' cp '{}' ../outfiles_scjack_540_TBE_processed/
```

There are 439 scjackknife trees remaining.

Moving forward, use the directories of `outtrees` and `outfiles` that have had errored runs removed for downstream analyses.

Removing files causes the index between the scjack matrices and `outtree` and `outfile` files to be mismatched, causing downstream analysis issues. Remove those scjack matrices that produced errored runs.

```{r id-errored-runs}
outtrees_dir = "~/repo/aqhumor/AqHum_5Species/aqhumor_final/scjackknife/scjack_54_jumble_search_TBE/scjack_run/output_scjack_540_TBE/outtrees_scjack_540_TBE/"
matrix_name_pattern = "scjackknife_540_TBE"

id_errored_runs(outtrees_dir = outtrees_dir, matrix_name_pattern = matrix_name_pattern)

```

The above function provides a list of matrices whose `contml` runs errored. Use this list to remove these matrices using bash:
```
cd scjack_matrices
rm -r $(<errored_runs.txt)

```
Moving forward, use the directory of matrices with error-causing matrices removed in downstream analysis.

# Calculate the scjack scores

## Prepare input files for Booster.

```{r boot-files}

#Prepare scjackknife trees
# scjackknife trees must share common labels in order to calculate the TBE. Label the scjackknife trees with their full cell ids and an abbreviated cell id that can be matched across scjackknife trees.

n_trees =439
matrix_dir_path = "~/repo/aqhumor/AqHum_5Species/aqhumor_final/scjackknife/scjack_54_jumble_search_TBE/scjack_run/output_scjack_540_TBE_processed/scjack_matrices_processed/"
tree_dir_path = "~/repo/aqhumor/AqHum_5Species/aqhumor_final/scjackknife/scjack_54_jumble_search_TBE/scjack_run/output_scjack_540_TBE_processed/outtrees_scjack_540_TBE_processed/"
file_name_prefix = "outtrees_scjack_540_TBE"

labelled_trees <- label_scjack_trees(n_trees = n_trees, matrix_dir_path = matrix_dir_path, tree_dir_path = tree_dir_path, print=TRUE, file_name_prefix=file_name_prefix)

#combine and print out as a single tree file
multi_trees <- combine_multi_trees("~/repo/aqhumor/AqHum_5Species/aqhumor_final/scjackknife/scjack_54_jumble_search_TBE/scjack_run/output_scjack_540_TBE_processed/outtrees_scjack_540_TBE_common_label/")

write.tree(multi_trees, file="multiphylo_scjack_540_TBE_common_label.tre")


#The phylogeny to calculate TBE scores for must also share common labels with the scjackknife trees.
tree <- read.tree("aqhumor_final/scjackknife/scjack_54_jumble_search_TBP/outtree_54_jumble_search_1249177875.tre")

labelled_tree <- label_tree(matrix_path = "aqhumor_final/jumble/54_jumble_search/contml_54cell_subset_pca_var_norm_20PC_mtx/", tree_path = "aqhumor_final/scjackknife/scjack_54_jumble_search_TBP/outtree_54_jumble_search_1249177875.tre", file_name = "outtree_54_jumble_search_1249177875_label.tre", print=TRUE)

#make new labels - code stolen from scjack label
species <- sapply(strsplit(labelled_tree$tip.label, "_"), function(v){return(v[1])})
cluster_id <- sapply(strsplit(labelled_tree$tip.label, "_"), function(v){return(v[3])})
new_label <- paste0(species, "_", cluster_id)
common_labels_tree <- labelled_tree
common_labels_tree$tip.label <- new_label
#print out common labels tree
write.tree(common_labels_tree, file="outtree_54_jumble_search_1249177875_common_label.tre")
    
```

## Run booster to calculate TBE scores
```
../../booster/booster_macos64 -i outtree_54_jumble_search_1249177875_common_label.tre -b multiphylo_scjack_540_TBE_common_label.tre  > outtree_54_jumble_search_1249177875_scored_with_scjack540TBE.tre
```

This produces a tree, `outtree_54_jumble_search_1249177875_scored_with_scjack540TBE.tre`, with TBE scores plotted at nodes. These TBE scores were used in Fig. 4 as the scjackknife scores.

